README.md
# http-simple-serve

> ⚙️ Simple HTTP server for use with static resources.

[![Test](https://github.com/neogeek/http-simple-serve/actions/workflows/test.workflow.yml/badge.svg)](https://github.com/neogeek/http-simple-serve/actions/workflows/test.workflow.yml)
[![Lint](https://github.com/neogeek/http-simple-serve/actions/workflows/lint.workflow.yml/badge.svg)](https://github.com/neogeek/http-simple-serve/actions/workflows/lint.workflow.yml)
[![Build](https://github.com/neogeek/http-simple-serve/actions/workflows/build.workflow.yml/badge.svg)](https://github.com/neogeek/http-simple-serve/actions/workflows/build.workflow.yml)
[![NPM Version](http://img.shields.io/npm/v/http-simple-serve.svg?style=flat)](https://www.npmjs.org/package/http-simple-serve)
[![Documentation](https://doxdox.org/images/badge-flat.svg)](https://doxdox.org/neogeek/http-simple-serve)

## Install

```bash
$ npm install http-simple-serve
```

## Usage

```javascript
import http from 'http-simple-serve';

http({
  port: process.env.PORT ? parseInt(process.env.PORT) : 8080,
  root: 'public/',
  entry: 'index.html',
});
```

## Custom Server

```javascript
import { readFileSync } from 'node:fs';
import { createServer } from 'node:http';

import { readStaticAssetsSync, tryParseUrl } from 'http-simple-serve';

const staticFiles = readStaticAssetsSync('./public', 'index.html');

createServer(async (req, res) => {
  try {
    const url = tryParseUrl(req.url);

    if (Object.keys(staticFiles).includes(url)) {
      const { path, contentType } = staticFiles[url];

      const content = readFileSync(path);

      res.setHeader('Content-type', contentType);

      res.writeHead(200);

      res.end(content);
    } else if (url === '/robots.txt') {
      res.writeHead(200);

      res.end('User-agent: *\nAllow: /');
    } else {
      res.writeHead(404);

      res.end('Not found');
    }
  } catch (error) {
    res.writeHead(500);

    res.end('Internal server error');
  }
}).listen(process.env.PORT ? parseInt(process.env.PORT) : 8080);
```

## Debugging

This will output all requests to the console with the path and response status code.

```bash
$ NODE_ENV=development npm start
```


CHANGELOG.md
# Changelog

## [v1.1.0](https://github.com/neogeek/http-simple-serve/tree/v1.1.0) - (2023-11-30)

[Full Changelog](https://github.com/neogeek/http-simple-serve/compare/v1.0.0...v1.1.0)

- [hotfix] Updated packages. [#4](https://github.com/neogeek/http-simple-serve/pull/4)
- [feat] Added auto entry path for sub directories. [#3](https://github.com/neogeek/http-simple-serve/pull/3)
- [hotfix] Added debug error log. [#2](https://github.com/neogeek/http-simple-serve/pull/2)
- [feat] Added tryParseUrl method. [#1](https://github.com/neogeek/http-simple-serve/pull/1)

## [v1.0.0](https://github.com/neogeek/http-simple-serve/tree/v1.0.0) - (2023-11-29)

- Initial release! 🎉

_This changelog was generated with **[generate-local-changelog](https://github.com/neogeek/generate-local-changelog)**_


src/index.test.ts
/* node:coverage disable */

import test from 'node:test';
import assert from 'node:assert';

import { readFileSync, unlinkSync, writeFileSync } from 'node:fs';

import http, {
  readStaticAssetsSync,
  tryGetContentType,
  tryParseUrl,
} from './index.js';

void test('http', async t => {
  await t.test(
    'start server and serve an entry file (indirectly)',
    async () => {
      const server = http({
        port: 0,
        root: './demo/public',
        entry: 'index.html',
      });

      const address = server.address();

      if (!address || typeof address !== 'object') {
        throw new Error('Server address not set correctly.');
      }

      const response = await fetch(`http://localhost:${address.port}`);

      assert.strictEqual(
        await response.text(),
        readFileSync('./demo/public/index.html', 'utf8')
      );

      server.close();
    }
  );
  await t.test('start server and serve an entry file (directly)', async () => {
    const server = http({
      port: 0,
      root: './demo/public',
      entry: 'index.html',
    });

    const address = server.address();

    if (!address || typeof address !== 'object') {
      throw new Error('Server address not set correctly.');
    }

    const response = await fetch(`http://localhost:${address.port}/index.html`);

    assert.strictEqual(
      await response.text(),
      readFileSync('./demo/public/index.html', 'utf8')
    );

    server.close();
  });
  await t.test('start server and serve a specific file', async () => {
    const server = http({
      port: 0,
      root: './demo/public',
      entry: 'index.html',
    });

    const address = server.address();

    if (!address || typeof address !== 'object') {
      throw new Error('Server address not set correctly.');
    }

    const response = await fetch(`http://localhost:${address.port}/index.html`);

    assert.strictEqual(
      await response.text(),
      readFileSync('./demo/public/index.html', 'utf8')
    );

    server.close();
  });
  await t.test('start server and serve a CSS file', async () => {
    const server = http({
      port: 0,
      root: './demo/public',
      entry: 'index.html',
    });

    const address = server.address();

    if (!address || typeof address !== 'object') {
      throw new Error('Server address not set correctly.');
    }

    const response = await fetch(
      `http://localhost:${address.port}/css/styles.css`
    );

    assert.strictEqual(
      await response.text(),
      readFileSync('./demo/public/css/styles.css', 'utf8')
    );

    server.close();
  });
  await t.test(
    'start server and serve a 404 for a invalid resource',
    async () => {
      const server = http({
        port: 0,
        root: './demo/public',
        entry: 'index.html',
      });

      const address = server.address();

      if (!address || typeof address !== 'object') {
        throw new Error('Server address not set correctly.');
      }

      const response = await fetch(
        `http://localhost:${address.port}/js/main.js`
      );

      assert.strictEqual(response.status, 404);

      server.close();
    }
  );
  await t.test(
    'start server and fail to serve file that gets deleted after the server starts',
    async () => {
      writeFileSync('./demo/public/500.html', '');

      const server = http({
        port: 0,
        root: './demo/public',
        entry: 'index.html',
      });

      unlinkSync('./demo/public/500.html');

      const address = server.address();

      if (!address || typeof address !== 'object') {
        throw new Error('Server address not set correctly.');
      }

      const response = await fetch(`http://localhost:${address.port}/500.html`);

      assert.strictEqual(response.status, 500);

      server.close();
    }
  );
});

void test('readStaticAssetsSync', async t => {
  await t.test('get files in directory', () => {
    assert.deepStrictEqual(
      readStaticAssetsSync('./demo/public', 'index.html'),
      {
        '/': {
          contentType: 'text/html; charset=utf-8',
          path: 'demo/public/index.html',
        },
        '/index.html': {
          contentType: 'text/html; charset=utf-8',
          path: 'demo/public/index.html',
        },
        '/css/styles.css': {
          contentType: 'text/css; charset=utf-8',
          path: 'demo/public/css/styles.css',
        },
        '/docs': {
          contentType: 'text/html; charset=utf-8',
          path: 'demo/public/docs/index.html',
        },
        '/docs/index.html': {
          contentType: 'text/html; charset=utf-8',
          path: 'demo/public/docs/index.html',
        },
      }
    );
  });
});

void test('tryGetContentType', async t => {
  await t.test('get content type from path', () => {
    assert.strictEqual(
      tryGetContentType('./public/index.html'),
      'text/html; charset=utf-8'
    );
  });
  await t.test('fail to get content type from invalid path', () => {
    assert.throws(
      () => {
        tryGetContentType('./public/');
      },
      {
        name: 'TypeError',
        message: 'Path specified was not valid.',
      }
    );
  });
});

void test('tryParseUrl', async t => {
  await t.test('get root url', () => {
    assert.strictEqual(tryParseUrl('/'), '/');
  });
  await t.test('get file path', () => {
    assert.strictEqual(tryParseUrl('/index.html'), '/index.html');
  });
  await t.test('get sub directory with trailing slash', () => {
    assert.strictEqual(tryParseUrl('/docs/'), '/docs');
  });
  await t.test('get sub directory without trailing slash', () => {
    assert.strictEqual(tryParseUrl('/docs'), '/docs');
  });
  await t.test('get file in sub directory', () => {
    assert.strictEqual(tryParseUrl('/docs/index.html'), '/docs/index.html');
  });
  await t.test('fail to parse url when url is missing', () => {
    assert.throws(
      () => {
        tryParseUrl(undefined);
      },
      {
        name: 'TypeError',
        message: 'URL is missing',
      }
    );
  });
});


demo/custom.js
/* eslint-env node */

import { readFileSync } from 'node:fs';
import { createServer } from 'node:http';

import { readStaticAssetsSync, tryParseUrl } from 'http-simple-serve';

const staticFiles = readStaticAssetsSync('./public', 'index.html');

createServer(async (req, res) => {
  try {
    const url = tryParseUrl(req.url);

    if (Object.keys(staticFiles).includes(url)) {
      const { path, contentType } = staticFiles[url];

      const content = readFileSync(path);

      res.setHeader('Content-type', contentType);

      res.writeHead(200);

      res.end(content);
    } else if (url === '/robots.txt') {
      res.writeHead(200);

      res.end('User-agent: *\nAllow: /');
    } else {
      res.writeHead(404);

      res.end('Not found');
    }
  } catch (error) {
    res.writeHead(500);

    res.end('Internal server error');
  }
}).listen(process.env.PORT ? parseInt(process.env.PORT) : 8080);


src/index.ts
import { basename, dirname, extname, join, normalize } from 'node:path';
import { lstatSync, readdirSync, readFileSync } from 'node:fs';
import { createServer } from 'node:http';

import { contentType } from 'mime-types';

/**
 * Returns a list of file paths and their content types found recursively in a given path.
 *
 * @example readStaticAssetsSync('./public', 'index.html');
 * @param {String} root Relative path to the root of a given directory.
 * @param {String} entry The default resource to return when accessing the root of the main directory.
 * @return {{[key: string]: { path: string; contentType: string; }}} The list of file paths and content types.
 */
export const readStaticAssetsSync = (root: string, entry: string) => {
  const staticFiles: {
    [key: string]: {
      path: string;
      contentType: string;
    };
  } = {} as const;

  readdirSync(root, { encoding: 'utf8', recursive: true }).map(path => {
    const stat = lstatSync(join(root, path));

    if (!stat.isFile()) {
      return;
    }

    if (basename(path) === entry) {
      staticFiles[normalize(`/${dirname(path)}`)] = {
        path: join(root, path),
        contentType: tryGetContentType(path),
      };
    }

    staticFiles[normalize(`/${path}`)] = {
      path: join(root, path),
      contentType: tryGetContentType(path),
    };
  });

  return staticFiles;
};

/**
 * Returns the content type of a file for use in an HTTP response header.
 *
 * @example tryGetContentType('./public/index.html');
 * @param {String} path A relative file path.
 * @return {String} The content type for the given file path.
 */
export const tryGetContentType = (path: string) => {
  const result = contentType(extname(path));

  if (!result) {
    throw TypeError('Path specified was not valid.');
  }

  return result;
};

/**
 * Returns a parsed URL. If the URL is to a directory, the trailing slash is removed.
 *
 * @example tryParseUrl('/docs/index.html');
 * @example tryParseUrl('/docs/');
 * @param {String} url The requested URL.
 * @return {String} The parsed URL.
 */
export const tryParseUrl = (url: string | undefined) => {
  if (!url) {
    throw TypeError('URL is missing');
  }

  if (url === '/') {
    return url;
  }

  return url.replace(/\/$/, '');
};

/**
 * Generates an HTTP server for serving up static resources.
 *
 * @example server({ port: 8080, root: './public', entry: 'index.html' });
 * @param {Object} options
 * @param {Number} options.port Port to run HTTP server from.
 * @param {String} options.root Relative path to the root of a given directory.
 * @param {String} options.entry The default resource to return when accessing the root of the main directory.
 * @return {Server<typeof IncomingMessage, typeof ServerResponse>} HTTP server.
 */
const server = ({
  port,
  root = './public',
  entry = 'index.html',
}: {
  port: number;
  root?: string;
  entry?: string;
}) => {
  const staticFiles = readStaticAssetsSync(root, entry);

  return createServer((req, res) => {
    try {
      const url = tryParseUrl(req.url);

      if (!Object.keys(staticFiles).includes(url)) {
        if (process.env.NODE_ENV === 'development') {
          console.log(`404 ${url}`);
        }

        res.writeHead(404);

        res.end('Not found');

        return;
      }

      const { path, contentType } = staticFiles[url];

      if (process.env.NODE_ENV === 'development') {
        console.log(`200 ${path} ${contentType}`);
      }

      const content = readFileSync(path);

      res.setHeader('Content-type', contentType);

      res.writeHead(200);

      res.end(content);
    } catch (error) {
      if (process.env.NODE_ENV === 'development') {
        console.log('500', error);
      }

      res.writeHead(500);

      res.end('Internal server error');
    }
  }).listen(port);
};

export default server;
